File: datasets/C++/Claude/DataStructures/hard5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>

using namespace std;

vector<int> bfs_distances(const vector<vector<int>>& graph, int start, int n) {
    vector<int> distances(n + 1, -1);
    distances[start] = 0;
    queue<int> q;
    q.push(start);
    
    while (!q.empty()) {
        int curr = q.front();
        q.pop();
        
        for (int neighbor : graph[curr]) {
            if (distances[neighbor] == -1) {
                distances[neighbor] = distances[curr] + 1;
                q.push(neighbor);
            }
        }
    }
    
    return distances;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    cin >> n >> m;
    
    // Build adjacency list for the tree
    vector<vector<int>> graph(n + 1);
    for (int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    
    // Precompute all pairwise distances in the tree
    vector<vector<int>> dist(n + 1, vector<int>(n + 1));
    for (int i = 1; i <= n; i++) {
        vector<int> distances = bfs_distances(graph, i, n);
        for (int j = 1; j <= n; j++) {
            dist[i][j] = distances[j];
        }
    }
    
    // Read the points in the squared metric space
    vector<pair<int, int>> points(m);
    for (int i = 0; i < m; i++) {
        cin >> points[i].first >> points[i].second;
    }
    
    // Find maximum distance between any two points
    int max_distance = 0;
    for (int i = 0; i < m; i++) {
        for (int j = i + 1; j < m; j++) {
            int x1 = points[i].first, y1 = points[i].second;
            int x2 = points[j].first, y2 = points[j].second;
            
            // Distance in squared metric space: ρ(x1,x2) + ρ(y1,y2)
            int distance = dist[x1][x2] + dist[y1][y2];
            max_distance = max(max_distance, distance);
        }
    }
    
    cout << max_distance << endl;
    
    return 0;
}

File: datasets/Python/Claude/DataStructures/hard5.py
from collections import deque, defaultdict

def bfs_distances(graph, start, n):
    """Calculate distances from start vertex to all other vertices using BFS"""
    distances = [-1] * (n + 1)
    distances[start] = 0
    queue = deque([start])
    
    while queue:
        curr = queue.popleft()
        for neighbor in graph[curr]:
            if distances[neighbor] == -1:
                distances[neighbor] = distances[curr] + 1
                queue.append(neighbor)
    
    return distances

def solve():
    # Read input
    n, m = map(int, input().split())
    
    # Build adjacency list for the tree
    graph = defaultdict(list)
    for _ in range(n - 1):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)
    
    # Precompute all pairwise distances in the tree
    dist = {}
    for i in range(1, n + 1):
        distances = bfs_distances(graph, i, n)
        for j in range(1, n + 1):
            dist[(i, j)] = distances[j]
    
    # Read the points in the squared metric space
    points = []
    for _ in range(m):
        x, y = map(int, input().split())
        points.append((x, y))
    
    # Find maximum distance between any two points
    max_distance = 0
    for i in range(m):
        for j in range(i + 1, m):
            x1, y1 = points[i]
            x2, y2 = points[j]
            
            # Distance in squared metric space: ρ(x1,x2) + ρ(y1,y2)
            distance = dist[(x1, x2)] + dist[(y1, y2)]
            max_distance = max(max_distance, distance)
    
    print(max_distance)

if __name__ == "__main__":
    solve()